File: 2015/06/1.py
input = None
try:
    input = open("./2015/06/input.txt", "r")
except:
    print("Error opening input.txt")
    exit(1)


class Lightgrid:
    lights = None

    def __init__(self) -> None:
        self.lights = []
        for _ in range(1000):
            _arr = []
            for _ in range(1000):
                _arr.append(False)
            self.lights.append(_arr)

    def on(self, begx, begy, endx, endy):
        for x in range(begx, endx + 1):
            for y in range(begy, endy + 1):
                self.lights[x][y] = True

    def off(self, begx, begy, endx, endy):
        for x in range(begx, endx + 1):
            for y in range(begy, endy + 1):
                self.lights[x][y] = False

    def toggle(self, begx, begy, endx, endy):
        for x in range(begx, endx + 1):
            for y in range(begy, endy + 1):
                self.lights[x][y] = not self.lights[x][y]

    def counton(self):
        count = 0
        for column in self.lights:
            for light in column:
                if light:
                    count += 1
        return count


nice_count = 0
line = None
lightgrid = Lightgrid()
while True:
    line = input.readline()
    if line == "":
        break
    # Trim away the ending newline character
    if line.endswith("\n"):
        line = line[:-1]

    fun = None
    if line.startswith("turn on "):
        line = line[len("turn on ") :]
        fun = lightgrid.on
    elif line.startswith("turn off "):
        line = line[len("turn off ") :]
        fun = lightgrid.off
    elif line.startswith("toggle "):
        line = line[len("toggle ") :]
        fun = lightgrid.off
        fun = lightgrid.toggle
    else:
        print("Error parsing input.txt unable to parse instruction")
        exit(1)

    [beg, end] = line.split(" through ")
    [begx, begy] = beg.split(",")
    [endx, endy] = end.split(",")
    begx = int(begx)
    begy = int(begy)
    endx = int(endx)
    endy = int(endy)
    fun(begx, begy, endx, endy)

print(lightgrid.counton())

File: 2015/06/1.js
"use strict";
const fs = require("node:fs");
let contents = fs.readFileSync("./2015/06/input.txt", "utf8").toString();

class Lightgrid {
  lights = undefined;
  constructor() {
    this.lights = [];
    for (let x = 0; x != 1000; x++) {
      this.lights[x] = [];
      for (let y = 0; y != 1000; y++) {
        this.lights[x][y] = false;
      }
    }
  }

  turnOn(begx, begy, endx, endy) {
    let stepx = begx < endx ? +1 : -1;
    let stepy = begy < endy ? +1 : -1;
    for (let x = begx; x != endx + stepx; x += stepx) {
      for (let y = begy; y != endy + stepy; y += stepy) {
        this.lights[x][y] = true;
      }
    }
  }
  turnOff(begx, begy, endx, endy) {
    let stepx = begx < endx ? +1 : -1;
    let stepy = begy < endy ? +1 : -1;
    for (let x = begx; x != endx + stepx; x += stepx) {
      for (let y = begy; y != endy + stepy; y += stepy) {
        this.lights[x][y] = false;
      }
    }
  }

  toggle(begx, begy, endx, endy) {
    let stepx = begx < endx ? +1 : -1;
    let stepy = begy < endy ? +1 : -1;
    for (let x = begx; x != endx + stepx; x += stepx) {
      for (let y = begy; y != endy + stepy; y += stepy) {
        this.lights[x][y] = !this.lights[x][y];
      }
    }
  }

  countOn() {
    let count = 0;
    for (let x = 0; x != 1000; x++) {
      for (let y = 0; y != 1000; y++) {
        if (this.lights[x][y]) {
          count++;
        }
      }
    }
    return count;
  }
}

let lightgrid = new Lightgrid();
for (const line of contents.split("\n")) {
  let { instruction, begx, begy, endx, endy } = line.match(
    /^(?<instruction>turn on|turn off|toggle) (?<begx>[0-9]+),(?<begy>[0-9]+) through (?<endx>[0-9]+),(?<endy>[0-9]+)$/,
  ).groups;
  begx = Number(begx);
  begy = Number(begy);
  endx = Number(endx);
  endy = Number(endy);
  switch (instruction) {
    case "turn on":
      lightgrid.turnOn(begx, begy, endx, endy);
      break;
    case "turn off":
      lightgrid.turnOff(begx, begy, endx, endy);
      break;
    case "toggle":
      lightgrid.toggle(begx, begy, endx, endy);
      break;
  }
}
console.log(lightgrid.countOn());
